db/context.go

```diff
@@ -0,0 +1,66 @@
+// Copyright 2024 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package db
+
+import (
+	"context"
+
+	"github.com/sourcenetwork/defradb/datastore"
+)
+
+// txnContextKey is the key type for transaction context values.
+type txnContextKey struct{}
+
+// explicitTxn is a transaction that is managed outside of a db operation.
+type explicitTxn struct {
+	datastore.Txn
+}
+
+func (t *explicitTxn) Commit(ctx context.Context) error {
+	return nil // do nothing
+}
+
+func (t *explicitTxn) Discard(ctx context.Context) {
+	// do nothing
+}
+
+// transactionDB is a db that can create transactions.
+type transactionDB interface {
+	NewTxn(context.Context, bool) (datastore.Txn, error)
+}
+
+// ensureContextTxn ensures that the returned context has a transaction.
+//
+// If a transactions exists on the context it will be made explicit,
+// otherwise a new implicit transaction will be created.
+func ensureContextTxn(ctx context.Context, db transactionDB, readOnly bool) (context.Context, datastore.Txn, error) {
+	txn, ok := TryGetContextTxn(ctx)
+	if ok {
+		return SetContextTxn(ctx, &explicitTxn{txn}), &explicitTxn{txn}, nil
+	}
+	txn, err := db.NewTxn(ctx, readOnly)
+	if err != nil {
+		return nil, txn, err
+	}
+	return SetContextTxn(ctx, txn), txn, nil
+}
+
+// TryGetContextTxn returns a transaction and a bool indicating if the
+// txn was retrieved from the given context.
+func TryGetContextTxn(ctx context.Context) (datastore.Txn, bool) {
+	txn, ok := ctx.Value(txnContextKey{}).(datastore.Txn)
+	return txn, ok
+}
+
+// SetContextTxn returns a new context with the txn value set.
+func SetContextTxn(ctx context.Context, txn datastore.Txn) context.Context {
+	return context.WithValue(ctx, txnContextKey{}, txn)
```

Comment:
The `ensureContextTxn` makes sure we don't overwrite an existing transaction. There is a case within `net/server.go` where we need the functionality to overwrite the transaction.

Reply:
not sure if having `ensureContextTxn` is enough in this case. `SetContextTxn` is a public method and can be called from anywhere. If `ensureContextTxn` is required to call before calling this one, why not to make it part of it? Otherwise I see no link between the two.

Reply:
I'm not seeing the issue here. The lifetime of the transaction is managed from the caller. If they overwrite a transaction and don't discard it, then the bug is in the caller's code.

Reply:
if the intention is to put this responsibility on the caller's side, it's fine, but then it needs to be documented.

Reply:
I've updated the function docs to be more clear.